### PR TITLE
[c10d] Start deprecating *_coalesced APIs

### PR DESCRIPTION
**Pull request by kwen2501**
- We consider that general users need not to use the `*_coalesced` APIs unless there is an extreme concern about performance.

- We are investigating using a context manager named `coalescing_manager` which wrap around multiple individual collectives to compose the coalescing hint, rather than giving each collective a *_coalesced variant.

### CODE REVIEW

**Review by XilunWu — APPROVED** [review]
lgtm

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1551,6 +1551,10 @@ def all_reduce_coalesced(tensors, op=ReduceOp.SUM, group=None, async_op=False):
         None, if not async_op or if not part of the group.
 
     """
+    warnings.warn(
+        "torch.distributed.all_reduce_coalesced will be deprecated. If you must "
+        "use it, please revisit our documentation later."
```
Comment (H-Huang) [inline torch/distributed/distributed_c10d.py:1556]: nit: link to docs?

```suggestion
        "use it, please revisit our documentation (https://pytorch.org/docs/master/distributed.html#collective-functions) later."
```

**Review by H-Huang — APPROVED** [review]
lgtm

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/distributed_c10d.py]: Good idea. Added!

**Pull request closed without merging**